src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java

```diff
@@ -207,61 +178,93 @@ public String getDisplayName() {
     }
 
     public void checkApiRateLimit(TaskListener listener, GitHub gitHub) throws IOException, InterruptedException {
-        taskListenerThreadLocal.set(listener);
-        setGitHub(gitHub);
+        setListener(listener);
         RateLimitChecker currentChecker = getChecker(gitHub.getApiUrl());
 
+        long count = 0;
+        while (checkApiRateLimitOnce(currentChecker, gitHub, count++)) {
+            // continue
+        }
+    }
+
+    /**
+     * For testing only
+     */
+    static boolean checkApiRateLimitOnce(RateLimitChecker currentChecker, GitHub gitHub, long count) throws IOException, InterruptedException {
         if (currentChecker instanceof RateLimitCheckerBase) {
             GHRateLimit rateLimit = gitHub.getRateLimit();
-            long count = 0;
-            while (((RateLimitCheckerBase)currentChecker).checkRateLimit(rateLimit.getCore(), count++)) {
-                rateLimit = gitHub.getRateLimit();
-            }
+            return ((RateLimitCheckerBase)currentChecker).checkRateLimit(rateLimit.getCore(), count);
         }
+        return false;
     }
 
     static abstract class RateLimitCheckerBase extends RateLimitChecker {
         @Override
         protected boolean checkRateLimit(GHRateLimit.Record rateLimitRecord,
                                          long count) throws InterruptedException {
+            if (count == 0) {
+                setExpiration(Long.MIN_VALUE);
+            }
+            long expiration = getExpiration();
+            long start = System.currentTimeMillis();
+            if (waitUntilRateLimit(start, expiration, count)) {
+                return true;
+            }
             return this.checkRateLimitImpl(rateLimitRecord, count);
         }
 
+        // internal for testing
         abstract boolean checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count) throws InterruptedException;
 
-        protected void waitUntilRateLimit(GHRateLimit.Record rateLimit, long expiration, long count) throws InterruptedException {
-            long now = System.currentTimeMillis();
+        static long calculateExpirationForMinimum(GHRateLimit.Record rateLimit, long now, long rateLimitResetMillis, int minimum) {
```

Comment:
the name is a bit confusing to me, minimum what, minimum rateLimit doesn't seem right to me if you include parameters when reading?

calculate expiration would possibly read better

Reply:
`calculateExpirationIfOverBudgetIntoBuffer`

Reply:
what about calculateExpiration, and put the quirks in the javadoc?

Reply:
It can't be `calculateExpiration` because it is only for the case where we're over budget into the buffer zone. 

Went with `calculateExpirationWhenBufferExceeded`